File: solutions/typescript/078-clone-graph.ts
/**
 * Problem 78: Clone Graph (LeetCode 133)
 * Difficulty: Med
 * Language: TypeScript
 */
type GNode = { val: number; neighbors: GNode[] };

function cloneGraph(node: GNode | null): GNode | null {
  const dfs = (n: GNode, visited: Map<number, GNode>): GNode => {
    if (visited.has(n.val)) return visited.get(n.val)!;
    const clone: GNode = { val: n.val, neighbors: [] };
    const newVisited = new Map(visited).set(n.val, clone);
    clone.neighbors = n.neighbors.map(nb => dfs(nb, newVisited));
    return clone;
  };
  return node ? dfs(node, new Map()) : null;
}

File: solutions/rust/078-clone-graph.rs
/**
 * Problem 78: Clone Graph (LeetCode 133)
 * Difficulty: Med
 * Language: Rust
 */
use std::collections::HashMap;
use std::rc::Rc;
use std::cell::RefCell;

fn clone_graph(node: Option<Rc<RefCell<GNode>>>,
               visited: &mut HashMap<i32, Rc<RefCell<GNode>>>)
  -> Option<Rc<RefCell<GNode>>> {
    node.map(|n| {
        let val = n.borrow().val;
        if let Some(c) = visited.get(&val) { return Rc::clone(c); }
        let clone = Rc::new(RefCell::new(GNode { val, neighbors: vec![] }));
        visited.insert(val, Rc::clone(&clone));
        let nbrs: Vec<_> = n.borrow().neighbors.iter()
            .filter_map(|nb| clone_graph(Some(Rc::clone(nb)), visited)).collect();
        clone.borrow_mut().neighbors = nbrs;
        clone
    })
}
